h5py/_conv.pyx

```diff
@@ -714,17 +716,38 @@ cdef int conv_vlen2ndarray(void* ipt,
     itemsize = H5Tget_size(outtype.id)
     if itemsize > H5Tget_size(intype.id):
         data = realloc(data, itemsize * size)
-    H5Tconvert(intype.id, outtype.id, size, data, NULL, H5P_DEFAULT)
+
+    if needs_bkg_buffer(intype.id, outtype.id):
+        back_buf = emalloc(H5Tget_size(outtype.id)*size)
+
+    try:
+        H5Tconvert(intype.id, outtype.id, size, data, back_buf, H5P_DEFAULT)
+    finally:
+        free(back_buf)
+
+    # We need to use different approaches to creating the ndarray with the converted
+    # data depending on the destination dtype.
+    # For simple dtypes, we can use SimpleNewFromData, but types like
+    # string & void need a size specified, so this function can't be used.
+    # Additionally, Cython doesn't expose NumPy C-API functions like NewFromDescr,
+    # so we fall back on copying directly to the underlying buffer
+    # of a new ndarray for other types.
 
     if elem_dtype.kind in b"biufcmMO":
         # type_num is enough to create an array for these dtypes
         ndarray = cnp.PyArray_SimpleNewFromData(1, dims, elem_dtype.type_num, data)
-    else:
-        # dtypes like string & void need a size specified, so can't be used with
-        # SimpleNewFromData. Cython doesn't expose NumPy C-API functions
+    elif not elem_dtype.hasobject:
+        # This covers things like string dtypes and simple compound dtypes,
+        # which can't be used with SimpleNewFromData.
+        # Cython doesn't expose NumPy C-API functions
         # like NewFromDescr, so we'll construct this with a Python function.
         buf = <char[:itemsize * size]> data
         ndarray = np.frombuffer(buf, dtype=elem_dtype)
+    else:
+        # Compound dtypes containing object fields: frombuffer() refuses these,
+        # so we'll fall back to allocating a new array and copying the data in.
+        ndarray = np.empty(size, dtype=elem_dtype)
+        memcpy(PyArray_DATA(ndarray), data, itemsize * size)
```

Comment:
One thing I'm not clear on: if we copy from `data` to a new location in memory, is our function responsible for freeing `data`? I'm guessing it is, because we already call `realloc(data, ...)` above, which can free the old pointer.

Reply:
@takluyver, I'm really sorry for letting this sit around for so long. I think you are right. HDF5 allocates the v-len buffer and we are responsible for freeing it when we are no longer using it, which is the case here.